core/shared/src/main/scala/scalaz/zio/IOQueue.scala

```diff
@@ -38,44 +34,73 @@ class IOQueue[A] private (capacity: Int, ref: IORef[State[A]]) {
    * the queue.
    */
   final def offer[E](a: A): IO[E, Unit] =
-    IO.flatten(ref.modifyFold[E, IO[E, Unit]] {
-      case Deficit(takers) =>
-        takers.dequeueOption match {
-          case None                  => (IO.unit[E], Surplus(Queue.empty[A].enqueue(a), Queue.empty))
-          case Some((taker, takers)) => (taker.complete[E](a).toUnit, Deficit(takers))
-        }
-      case Surplus(values, putters) =>
-        if (values.length < capacity && putters.isEmpty) {
-          (IO.unit[E], Surplus(values.enqueue(a), putters))
-        } else {
-          val p = Promise.unsafeMake[E, Unit]
-          (p.get.ensuring(removePutter(p)), Surplus(values, putters.enqueue((a, p))))
-        }
-    })
+    for {
+      pRef <- IORef[E, Option[Promise[E, Unit]]](None)
+      a <- (for {
+            p <- ref
+                  .modifyFold[Void, (Promise[E, Unit], IO[Void, Unit])] {
+                    case Deficit(takers) =>
+                      val p = Promise.unsafeMake[E, Unit]
+
+                      takers.dequeueOption match {
+                        case None                  => ((p, p.complete(()).toUnit), Surplus(Queue.empty[A].enqueue(a), Queue.empty))
+                        case Some((taker, takers)) => ((p, taker.complete[Void](a) *> p.complete[Void](()).toUnit), Deficit(takers))
+                      }
+                    case Surplus(values, putters) =>
+                      val p = Promise.unsafeMake[E, Unit]
+
+                      if (values.length < capacity && putters.isEmpty) {
+                        ((p, p.complete(()).toUnit), Surplus(values.enqueue(a), putters))
+                      } else {
+                        val p = Promise.unsafeMake[E, Unit]
+                        ((p, IO.unit[Void]), Surplus(values, putters.enqueue((a, p))))
+                      }
+                  }
+                  .flatMap(t => t._2 *> pRef.write(Some(t._1)) *> IO.now(t._1))
```

Comment:
The four cases can be summarised:
```scala
(p, p.complete(()).toUnit)
(p, taker.complete[Void](a) *> p.complete[Void](()).toUnit)
(p, p.complete(()).toUnit)
(p, IO.unit[Void])
```
It looks like `p` is only useful in the last case and the IO action in the second case. I'd suggest this instead:
```scala
None
Some(Left(taker.complete[Void](a))
None
Some(Right(p))
```
And then this action becomes
```scala
_.fold(
  IO.now(Promise.done(())),
  Left(t) => t.map(_ => Promise.done(()))
  Right(p) => pRef.write(Some(p)).map(_ => p)
)
```

Reply:
(assuming the existence of `Promise.done` which lifts a pure value in a resolved promise)

Reply:
Some refactoring can be done here. The important thing is the promise `p` must exist and its value must be written into `pRef` in an uninterruptible block tied to the `modifyFold`.